netket/experimental/sampler/rules/fermion_2nd.py

```diff
@@ -0,0 +1,125 @@
+import jax
+import jax.numpy as jnp
+from typing import Optional
+import numpy as np
+
+from netket.sampler.rules import ExchangeRule
+from netket.graph import AbstractGraph
+from netket.graph import disjoint_union
+from netket.experimental.hilbert import SpinOrbitalFermions
+
+
+class ParticleExchangeRule(ExchangeRule):
+    """Exchange rule for particles on a lattice.
+
+    Works similarly to exchange rule, but:
+    takes into account that only occupied orbitals
+    can be exchanged with unoccupied ones.
+    This sampler conserves the number of particles.
+    """
+
+    copy_per_spin: bool
+    """
+    Whether to copy the graph for each spin sector.
+    This only occurs if cluster indices or number of graph nodes
+    are lower than the number of orbitals per spin sector.
+    """
+
+    def __init__(
+        self,
+        hilbert,
+        *,
+        clusters: Optional[list[tuple[int, int]]] = None,
+        graph: Optional[AbstractGraph] = None,
+        d_max: int = 1,
+        copy_per_spin: bool = True,
```

Comment:
Actually @jwnys , would it make sense to default this to `None` and have it default to
```python
if copy_per_spin is None:
    copy_per_spin = all(n is not None for n in hilbert.n_fermions_per_spin)
```
?